Make diff of credentials more readable

sinsoku opened the pull request:
### Summary

The `credentials.yml.enc` is encrypted, so you can't directly read diff.

```diff
$ git diff config/credentials.yml.enc
diff --git a/config/credentials.yml.enc b/config/credentials.yml.enc
index cbde4b6..33d24bb 100644
--- a/config/credentials.yml.enc
+++ b/config/credentials.yml.enc
@@ -1 +1 @@
-SkA3xhP9vZocyRxcvUu9caHkTZB9kMsQlSXT7JUOmcENs0zbKBaGysBu3S9d10RU1E8vWNAjFOkH+L/ZcOaW17tBRj6XpNL0Vn7C7VWXRITMpWhuhjqliAHOFn6NuR+6vXvl5IqlSnu3Km7TmsdxKKNNeViHUB3r8lcRg1HCW6j5rHZwQNQzvO+ZTVNqdtZIWx/FDEBc91ErgqXY+fvl6/nedyhQjGlGBEkQ9SyUDHYSuVwqps43fUvocUktjsV7Yy4Yoo6yhdq65jmNO6xs9ANc9aVwIBzJLk97KXEqOBDOk2UBFMDOnVj+lcsEA1iC2prta1txe0uKR4YZGa/xVc5jsoG/TaKKpJG02Zozn/Xo695PQRDd3YQ+KeIhsmX3w+heXJeQubQdHV5Sndq8uGfe54ubAW6cM1vh--SzZNsV3nxEA+22ff--XVmpjrbh6BBAtmwgsSh45g==
\ No newline at end of file
+RRPKRwBRwxmMvAG/z2M+rCUbTIi+1MyJM9shdeeuFnR8j5pK/a0H71WUU4BwTrwYleK5HTQO44EtJqWprpxP4FR53WXnF1kkdHAzk1qZdr/xhGRv7ChdZ9SV/M4uddIOjndc8HTdnAhxvGEa3DnwZm0SyHrXNDujUMbDE4sWEqrdSCTlUs+dOcHRnhbMbtRliKE3szlTIoQ2MRb9+37DkukKBYFtQWfDkStMT3VZ4VB/IvN2XQ6b685mQD+ncxVbWjBQ3nLmv7Y0hMbZirAIrfUIjAeEK6lzMQhhAauMI3juXa4h2ppyY5J+xmt5OAH+OwBi/ojaroaQ8SDkb87JMgztf3VGE6/2/6xF9yZthkSIdLODfII0PpztZXg0CM+XUZsIRoZTvPaUhLvybNi7JvKSKueFjGpO0epBm2F04/ay+LRZ--bedrdlaBiIITj7yF--pAO5A0jMR4RZl7S0RA784A==
\ No newline at end of file
```

However, if you tell Git how to decrypt files using Git attributes, you will be able to compare via the normal diff.

```diff
$ git diff config/credentials.yml.enc
diff --git a/config/credentials.yml.enc b/config/credentials.yml.enc
index 9413953..794e833 100644
--- a/config/credentials.yml.enc
+++ b/config/credentials.yml.enc
@@ -4,3 +4,4 @@

 # Used as the base secret for all MessageVerifiers in Rails, including the one protecting cookies.
 secret_key_base: 2c0c93402b7024b8fc00790cb8f496211cd61e61cdf39faa44a263bb5d573cded2b00a0c513227d5aaa314ebb1b40ea23566604bb741e68229f2f880caa2bf7a
+api_token: 123
```

This commit adds settings to Git so that you can display diff when creating a new app.

### Other Information

- https://git-scm.com/book/en/v2/Customizing-Git-Git-Attributes

sinsoku committed:
Make diff of credentials more readable

The `credentials.yml.enc` is encrypted, so you can't directly read diff.

```diff
$ git diff config/credentials.yml.enc
diff --git a/config/credentials.yml.enc b/config/credentials.yml.enc
index cbde4b6..33d24bb 100644
--- a/config/credentials.yml.enc
+++ b/config/credentials.yml.enc
@@ -1 +1 @@
-SkA3xhP9vZocyRxcvUu9caHkTZB9kMsQlSXT7JUOmcENs0zbKBaGysBu3S9d10RU1E8vWNAjFOkH+L/ZcOaW17tBRj6XpNL0Vn7C7VWXRITMpWhuhjqliAHOFn6NuR+6vXvl5IqlSnu3Km7TmsdxKKNNeViHUB3r8lcRg1HCW6j5rHZwQNQzvO+ZTVNqdtZIWx/FDEBc91ErgqXY+fvl6/nedyhQjGlGBEkQ9SyUDHYSuVwqps43fUvocUktjsV7Yy4Yoo6yhdq65jmNO6xs9ANc9aVwIBzJLk97KXEqOBDOk2UBFMDOnVj+lcsEA1iC2prta1txe0uKR4YZGa/xVc5jsoG/TaKKpJG02Zozn/Xo695PQRDd3YQ+KeIhsmX3w+heXJeQubQdHV5Sndq8uGfe54ubAW6cM1vh--SzZNsV3nxEA+22ff--XVmpjrbh6BBAtmwgsSh45g==
\ No newline at end of file
+RRPKRwBRwxmMvAG/z2M+rCUbTIi+1MyJM9shdeeuFnR8j5pK/a0H71WUU4BwTrwYleK5HTQO44EtJqWprpxP4FR53WXnF1kkdHAzk1qZdr/xhGRv7ChdZ9SV/M4uddIOjndc8HTdnAhxvGEa3DnwZm0SyHrXNDujUMbDE4sWEqrdSCTlUs+dOcHRnhbMbtRliKE3szlTIoQ2MRb9+37DkukKBYFtQWfDkStMT3VZ4VB/IvN2XQ6b685mQD+ncxVbWjBQ3nLmv7Y0hMbZirAIrfUIjAeEK6lzMQhhAauMI3juXa4h2ppyY5J+xmt5OAH+OwBi/ojaroaQ8SDkb87JMgztf3VGE6/2/6xF9yZthkSIdLODfII0PpztZXg0CM+XUZsIRoZTvPaUhLvybNi7JvKSKueFjGpO0epBm2F04/ay+LRZ--bedrdlaBiIITj7yF--pAO5A0jMR4RZl7S0RA784A==
\ No newline at end of file
```

However, if you tell Git how to decrypt files using Git attributes,
you will be able to compare via the normal diff.

```diff
$ git diff config/credentials.yml.enc
diff --git a/config/credentials.yml.enc b/config/credentials.yml.enc
index 9413953..794e833 100644
--- a/config/credentials.yml.enc
+++ b/config/credentials.yml.enc
@@ -4,3 +4,4 @@

 # Used as the base secret for all MessageVerifiers in Rails, including the one protecting cookies.
 secret_key_base: 2c0c93402b7024b8fc00790cb8f496211cd61e61cdf39faa44a263bb5d573cded2b00a0c513227d5aaa314ebb1b40ea23566604bb741e68229f2f880caa2bf7a
+api_token: 123
```

This commit adds settings to Git so that you can display diff when creating a new app.

matthewd commented on railties/lib/rails/generators/rails/app/app_generator.rb:
```diff
@@ -63,13 +63,15 @@ def configru
     def version_control
       if !options[:skip_git] && !options[:pretend]
         run "git init", capture: options[:quiet]
+        run "git config diff.rails_credentials.textconv 'bin/rails credentials:show'", capture: options[:quiet]
```
This only helps the first person to create the repository... maybe `bin/setup` would be better?

y-yagi replied:
+1 I think that `bin/setup` is a more appropriate place. Also, since, added it to `bin/setup`, I think that it is not necessary to run this with rails new.

matthewd commented on railties/lib/rails/commands/credentials/credentials_command.rb:
```diff
@@ -30,10 +30,11 @@ def edit
       end
 
-      def show
+      def show(path = nil, *)
```
Why the splat? Particularly if we support a specific positional parameter, it seems important to complain about unexpected extras.

matthewd replied:
Is this worth the positional parameter? Maybe it'd be better as an option.

matthewd reviewed (commented):
Interesting idea!

How does this behave if the key is unavailable?

sinsoku committed:
Use an option instead of the positional parameter

sinsoku committed:
Add support for people who clone a repo

Since it is necessary to configure the setting in the local repo each person, so it adds a command.

sinsoku commented:
> How does this behave if the key is unavailable?

It does not display the diff for `config/credentials.yml.enc` if the key is unavailable.
Because the `credentials:show` displays an error message when the key is unavailable and to generate the diff from this same two messages.

y-yagi commented on railties/lib/rails/generators/rails/app/templates/bin/setup.tt:
```diff
@@ -30,6 +30,11 @@ chdir APP_ROOT do
+<% unless options.skip_git? -%>
+
+  puts "\n== Configuring Git for Credentials =="
+  system! "git config diff.rails_credentials.textconv 'bin/rails credentials:show'"
```
Is it correct that no `path` option is specified?

kaspth commented on railties/lib/rails/commands/credentials/credentials_command.rb:
```diff
@@ -33,7 +35,8 @@ def edit
         require_application_and_environment!
 
-        say Rails.application.credentials.read.presence || missing_credentials_message
+        credentials = options.path? ? Rails.application.encrypted(options.path) : Rails.application.credentials
```
Think the path option belongs in the encrypted command, no?

kaspth commented on railties/lib/rails/generators/app_base.rb:
```diff
@@ -38,7 +38,7 @@ def self.add_shared_options_for(name)
 
         class_option :skip_git,            type: :boolean, aliases: "-G", default: false,
-                                           desc: "Skip .gitignore file"
+                                           desc: "Skip Git files"
```
Let's say "Git related"

kaspth reviewed (commented):
Nice idea!

schneems commented:
No response in about 4 months, you still interested in getting this in?

Edouard-chin commented:
If you are ok @sinsoku , I will pick this up unless you have plan to work on it. I'd love to see this shipped for Rails 6.0, it's a great idea and would be a shame if we forget it :)

paracycle commented:
It would also be great if this handled all `.yml.enc` files under `config/`, instead of just `config/credentials.yml.enc`. That way, we can support the same behaviour for multi-environment credentials added via #33521. We just need to be careful to not match `config/secrets.yml.enc` while doing that though, or to include support for that as well.

Edouard-chin commented:
> If you are ok @sinsoku , I will pick this up unless you have plan to work on it. I'd love to see this shipped for Rails 6.0, it's a great idea and would be a shame if we forget it :)

Sorry for the delay, time flies I guess 😅 . I'm probably too late for 6.0 but I opened the PR for review here https://github.com/rails/rails/pull/36777

sinsoku commented:
@Edouard-chin Thank you for picking this up.
I think your implementation of creating files when user types `credentials:edit` is awesome.

rafaelfranca commented:
Thank you @sinsoku

jensljungblad commented:
> It would also be great if this handled all `.yml.enc` files under `config/`, instead of just `config/credentials.yml.enc`. That way, we can support the same behaviour for multi-environment credentials added via #33521. We just need to be careful to not match `config/secrets.yml.enc` while doing that though, or to include support for that as well.

@paracycle Was this addressed? Trying to figure out how to diff environment specific credentials..
